### PR TITLE
Psycopg3 sync and async instrumentation  

### PR DESCRIPTION
**Pull request by reiktar**
# Description

Add psycopg3 Instrumentation including Instrumentation for psycopg3 Asynchronous connection support. The work is derived from psycopg2

given psycopg2 imlementation as foundation This PR 
 - adds psycopg3 support 
 - adds asynchronous connection Instrumentation
 - adds tests to cover the new asynchronous connections and cursors 
 
Fixes [#1751](https://github.com/open-telemetry/opentelemetry-python-contrib/issues/1751)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Expands the test coverage to include asynchronous connections
- [x] Tests are runnable via tox
- [x] Code is running on internal system in my organization without any known issues. 

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [x] Documentation has been updated

### CODE REVIEW

**Comment by srikanthccv** [comment]
There is already existing PR https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2123 for psycopg3

**Comment by reiktar** [comment]
> There is already existing PR #2123 for psycopg3

After reviewing #2123 the notable difference is instrumentation for Asynchronous connections. This branch/pr contains Instrumentation for all the same changes as #2123 introduces but also extends open-telemetry support for psycopg3 async connections and cursors and test-cases for the asynchronous instrumentation

**Comment by federicobond** [comment]
This is great work @reiktar! My initial goal with #2123 was to get _something_ out the door that could be improved later.

I think it may still make sense to merge #2123 first as it's an almost copy of the existing psycopg2 implementation, then rebase async support on top of it, so reviewing the async code is easier for the maintainers.

**Comment by reiktar** [comment]
@federicobond: Sorry for the late reply.
I see no problem with #2123 going first. I'd eventually like asynchronous psycopg3 support in open telemetry, as my organization relies heavily on async psycopg3. 

We actively use the async solution in all of our test and production environments but also wanted to contribute to the community.

**Comment by federicobond** [comment]
Awesome, @ocelotl has asked me to become component owner for the psycopg instrumentation before merging my PR. I promise to review your async code in a timely manner once we get the base support merged in.

**Comment by federicobond** [comment]
Hi @reiktar! Can I ask you to rebase your contribution on top of latest `main`?

**Comment by reiktar** [comment]
@federicobond: I have Updated the PR to expand on the latest main.

**Review by federicobond — COMMENTED** [review]
Looks good @reiktar! Left some small comments and questions.

File: tox.ini
Comment (reiktar) [reply tox.ini]: There were an issue that i ran into with regards to the Wrapt library. when the C extensions were turned on  ran into the following issue with `test_wrap_async_connection_class_with_cursor`.

```python
    async def test_async_connection():
>       acnx = await psycopg.AsyncConnection.connect(database="test")
E       AttributeError: 'method' object has no attribute '__get__'

test_psycopg_integration.py:229: AttributeError
```

Did another test right now and this is simply a problem of a missing `async` in the test case itself.

**Review by federicobond — APPROVED** [review]

File: CHANGELOG.md
Comment (xrmx) [inline CHANGELOG.md:31]: This should be moved to Unreleased  ↑

**Review by xrmx — APPROVED** [review]
LGTM, please move the changelog to Unreleased

**Review by aabmass — APPROVED** [review]
Didn't look too closely at the tests but LGTM

**Comment by reiktar** [comment]
@aabmass , @xrmx : There were two failed checks
    - One check seems to be unrelated to the changes at hand. 
    - The remaining check was pylint of the relevant psycopg code

What's confusing with that change set is that pylint also complains about existing/untouched psycopg code. In any case should i have a change set resolving the pylint issues ready to go should it be wished for. The PR has gone through several approvals

Just let me know and I will commit updates for  remaining pylint issues.

**Comment by reiktar** [comment]
What remains seems to be a conflict between black formatting and pylint.

**Comment by xrmx** [comment]
Last run has isort complaining in https://github.com/open-telemetry/opentelemetry-python-contrib/actions/runs/8300772936/job/22719403342?pr=2146#step:7:3359:

```
--- /home/runner/work/opentelemetry-python-contrib/opentelemetry-python-contrib/instrumentation/opentelemetry-instrumentation-psycopg/src/opentelemetry/instrumentation/psycopg/__init__.py:before	2024-03-15 18:29:36.777932
+++ /home/runner/work/opentelemetry-python-contrib/opentelemetry-python-contrib/instrumentation/opentelemetry-instrumentation-psycopg/src/opentelemetry/instrumentation/psycopg/__init__.py:after	2024-03-15 18:35:53.815141
@@ -106,11 +106,11 @@
 from typing import Collection
 
 import psycopg  # pylint: disable=import-self
-from psycopg import (  # pylint: disable=import-self,no-name-in-module
-    AsyncCursor as pg_async_cursor,
+from psycopg import (
+    AsyncCursor as pg_async_cursor,  # pylint: disable=import-self,no-name-in-module
 )
-from psycopg import (  # pylint: disable=no-name-in-module,import-self
-    Cursor as pg_cursor,
+from psycopg import (
+    Cursor as pg_cursor,  # pylint: disable=no-name-in-module,import-self
 )
 from psycopg.sql import Composed  # pylint: disable=no-name-in-module

 ```

**Review by ocelotl — APPROVED** [review]

**Comment by lzchen** [comment]
@reiktar 

Any interest in becoming a component owner as well for this instrumentation?

**Comment by reiktar** [comment]
@lzchen: I'm open to helping out, let me know what i need to do to get onboard.

**Pull request merged**
